### PR TITLE
Storage cleanup

### PR DESCRIPTION
**Pull request by monstermunchkin**

### CODE REVIEW

File: lxd/logging.go [hunk 1]
```diff
@@ -115,7 +115,7 @@ func expireLogs(ctx context.Context, state *state.State) error {
 				if logfile.IsDir() {
 					newest := newestFile(path, logfile)
 					if time.Since(newest).Hours() >= 48 {
-						os.RemoveAll(path)
+						err := os.RemoveAll(path)
```
Comment (stgraber) [inline lxd/logging.go]: This looks like something that should be its own commit (and probably be sent separately as a bugfix).

**Comment by stgraber** [comment]
@monstermunchkin can you split that bugfix into its own commit, then send a separate PR which includes that as well as the first 6 commits of this branch?

Those all seem non-controversial cleanup/fixes that we can very quickly land, reducing the size of this PR quite a bit by having it focus on moving code to the new package and then porting `dir` to the new logic.

File: lxd/storage_utils.go [hunk 2]
```diff
@@ -35,7 +36,7 @@ var MountOptions = map[string]mountOptions{
 	"diratime":      {false, unix.MS_NODIRATIME},
 	"dirsync":       {true, unix.MS_DIRSYNC},
 	"exec":          {false, unix.MS_NOEXEC},
-	"lazytime":      {true, MS_LAZYTIME},
+	"lazytime":      {true, driver.MS_LAZYTIME},
```
Comment (stgraber) [inline lxd/storage_utils.go]: This is appears to be defined in the `unix` package nowadays so we can switch that to unix.MS_LAZYTIME and drop the special logic in cgo.

File: lxd/storage/utils.go [hunk 3]
```diff
@@ -0,0 +1,406 @@
...
+// /var/lib/lxd/[snapshots|containers]/name
+func containerPath(projectName string, name string, isSnapshot bool) string {
+	if isSnapshot {
+		return shared.VarPath("snapshots", project.Prefix(projectName, name))
+	}
+
+	return shared.VarPath("containers", project.Prefix(projectName, name))
+}
+
+func setUnprivUserACL(idmapset *idmap.IdmapSet, destPath string) error {
```
Comment (stgraber) [inline lxd/storage/utils.go]: Where does that come from? I can't find it in the current code.

File: lxd/storage/utils.go [hunk 4]
```diff
@@ -0,0 +1,406 @@
...
+	fields := strings.SplitN(name, shared.SnapshotDelimiter, 2)
+	if len(fields) == 1 {
+		return name, "", false
+	}
+
+	return fields[0], fields[1], true
+}
+
+// /var/lib/lxd/[snapshots|containers]/name
+func containerPath(projectName string, name string, isSnapshot bool) string {
```
Comment (stgraber) [inline lxd/storage/utils.go]: Defined but not used in this commit

File: lxd/storage/utils.go [hunk 5]
```diff
@@ -0,0 +1,406 @@
...
+	}
+
+	err := shrinkFileSystem(fsType, devPath, mntpoint, byteSize)
+	return cleanupFunc, err
+}
+*/
+
+// Returns the parent container name, snapshot name, and whether it actually was
+// a snapshot name.
+func containerGetParentAndSnapshotName(name string) (string, string, bool) {
```
Comment (stgraber) [inline lxd/storage/utils.go]: Defined but not used in this commit

File: lxd/storage/utils.go [hunk 6]
```diff
@@ -0,0 +1,406 @@
...
+		}
+	default:
+		return fmt.Errorf(`Shrinking not supported for filesystem type "%s"`, fsType)
+	}
+
+	return nil
+}
+
+/*
+func shrinkVolumeFilesystem(s StorageDriver, volumeType int, fsType string, devPath string, mntpoint string, byteSize int64, data interface{}) (func() (bool, error), error) {
```
Comment (stgraber) [inline lxd/storage/utils.go]: Why is this all commented?

File: lxd/storage/utils.go [hunk 7]
```diff
@@ -0,0 +1,406 @@
...
+		errorMsg := fmt.Sprintf(`Could not extend underlying %s filesystem for "%s": %s`, fsType, devPath, msg)
+		logger.Errorf(errorMsg)
+		return fmt.Errorf(errorMsg)
+	}
+
+	logger.Debugf(`extended underlying %s filesystem for "%s"`, fsType, devPath)
+	return nil
+}
+
+func shrinkFileSystem(fsType string, devPath string, mntpoint string, byteSize int64) error {
```
Comment (stgraber) [inline lxd/storage/utils.go]: Defined but not used in commit

File: lxd/storage/utils.go [hunk 8]
```diff
@@ -0,0 +1,406 @@
...
+		"-u",
+		lvpath)
+	if err != nil {
+		return msg, err
+	}
+
+	return "", nil
+}
+
+func growFileSystem(fsType string, devPath string, mntpoint string) error {
```
Comment (stgraber) [inline lxd/storage/utils.go]: Defined but not used in commit

**Comment by stgraber** [comment]
Left a few comments on that second commit, then stopped as I'm really not sure what's going on in there. It felt like the goal was to move some of the utils functions into the new package, which I agree is a great first step (and is what we've been doing with @tomponline on the device package), but this commit on its own seems to move a lot of functions that aren't then used or removed from anywhere.

It could be that the rest of it comes from later commits and that it's just some code reshuffling that needs to happen here.

Ideally I'd have that commit lay some groundwork for the next steps by moving the needed functions away from the main lxd package into the storage package, make the needed ones public, give them a proper description (`golint` should be happy with the storage package) and update the references in the legacy code to import the storage package and use those exported functions.

**Comment by brauner** [comment]
This needs a rebase. :)

File: lxd/storage_utils.go (hunk 2 above)
Comment (monstermunchkin) [reply lxd/storage_utils.go]: This has been fixed upstream.

File: lxd/storage/utils.go (hunk 3 above)
Comment (monstermunchkin) [reply lxd/storage/utils.go]: This is a left over function which is removed again in the following commit. I fixed the commits so that it doesn't appear at all.

File: lxd/storage/utils.go (hunk 4 above)
Comment (monstermunchkin) [reply lxd/storage/utils.go]: I created a commit which moves this function to shared.

File: lxd/storage/utils.go (hunk 6 above)
Comment (monstermunchkin) [reply lxd/storage/utils.go]: Fixed

File: lxd/storage/utils.go (hunk 7 above)
Comment (monstermunchkin) [reply lxd/storage/utils.go]: This is used for other backends which originally were part of this PR.

File: lxd/storage/utils.go (hunk 8 above)
Comment (monstermunchkin) [reply lxd/storage/utils.go]: Again, used by other backends.

**Comment by stgraber** [comment]
@monstermunchkin 

As we discussed yesterday, please send a separate PR which:
 - Moves the cgo stuff to the separate package
 - Moves (rather than copy) the utils functions across too
 - Export those functions we need from the main package
 - Update the function calls to use the new package
 - Move ContainerPath to the new package too
 - Add new package to the golint test

With that branch done and merge this branch should just have the remaining 4 commits that are directly related to the new interface and dir backend.

**Comment by stgraber** [comment]
@monstermunchkin rebase please

**Comment by stgraber** [comment]
@monstermunchkin more rebasing :)

**Comment by stgraber** [comment]
Ok, so at first glance, this does achieve the goal of moving logic into a separate `storage` package, though we're not actually using Go interfaces so there is still some structural rework to be done.

Ideally, I'd like to get to something similar to our newly minted `device` package.

That is, have a `StorageDriver` interface which is exported by the storage package.
That interface is the only thing that will really be accessible to the outside world, so we need to make sure that all its exported functions make sense for all drivers and also fit the expected workflow.

The `Dir` struct then has to implement that interface.

And we then setup a single entrypoint which lets you load a particular storage driver by its name (`dir` in this case), passing and validating the minimal amount of configuration needed by storage drivers.

With that setup, `Dir` wouldn't be exported at all, it will have public functions for sure to match the public interface, but it will not itself be callable directly, forcing all callers through the single entry point and forcing all callers to treat all drivers as implementations of StorageDriver without being able to differentiate between them.

**Comment by stgraber** [comment]
I'm hoping I'll find some more time this weekend to poke at this and come up with more concrete suggestions on how to restructure the logic. As the package split itself looks good, the vast majority of what I'm suggesting will just boil down to tweaking what's exported and what's not and then bikesheeding on what's part of the StorageDriver interface and why.

**Comment by monstermunchkin** [comment]
@stgraber could you please review?

**Comment by tomponline** [comment]
jenkins: test this please

File: lxd/storage/dir.go [hunk 9]
```diff
@@ -0,0 +1,532 @@
...
+		return nil
+	}
+
+	return os.Rename(oldPath, newPath)
+}
+
+// VolumeMount mounts the given storage volume.
+func (s *dir) VolumeMount(project string, name string, volumeType VolumeType) (bool, error) {
+	// Nothing to do
+	return true, nil
```
Comment (tomponline) [inline lxd/storage/dir.go]: @monstermunchkin why does this function return both a bool and a err? Can it be modified such that a nil-err indicates success? (I may be missing something here :))

File: lxd/storage/dir_utils.go [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+
+import (
+	"os"
+
+	"github.com/lxc/lxd/lxd/project"
+	"github.com/lxc/lxd/shared"
+)
+
+// DirSnapshotDeleteInternal deletes the given snapshot internally.
+func DirSnapshotDeleteInternal(projectName, poolName string, snapshotName string) error {
```
Comment (tomponline) [inline lxd/storage/dir_utils.go]: @monstermunchkin can this function be made unexported from the `storage` package if it is internal?

File: lxd/storage/common.go [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+	d.pool = pool
+	d.poolID = poolID
+	d.volume = volume
+}
+
+func (d *driverCommon) GetVersion() string {
+	return d.sTypeVersion
+}
+
+func (d *driverCommon) UsesThinpool() bool {
```
Comment (tomponline) [inline lxd/storage/common.go]: @monstermunchkin this seems like something specific to LVM, is it really shared for all storage engines?

File: lxd/storage/dir.go [hunk 12]
```diff
@@ -0,0 +1,532 @@
...
+}
+
+// VolumeUmount unmounts the given storage volume.
+func (s *dir) VolumeUmount(project string, name string, volumeType VolumeType) (bool, error) {
+	// Nothing to do
+	return true, nil
+}
+
+// VolumeGetUsage returns the usage of the given storage volume.
+func (s *dir) VolumeGetUsage(project, name string,
```
Comment (tomponline) [inline lxd/storage/dir.go]: @monstermunchkin just a nit, but can this be moved onto 1 line?

File: lxd/storage/dir.go [hunk 13]
```diff
@@ -0,0 +1,532 @@
...
+	path string) (int64, error) {
+	return -1, fmt.Errorf("The directory container backend doesn't support quotas")
+}
+
+// VolumeSetQuota sets quotas for the given storage volume.
+func (s *dir) VolumeSetQuota(project, name string,
```
Comment (tomponline) [inline lxd/storage/dir.go]: @monstermunchkin same with this one.

File: lxd/storage/lock.go [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+	return fmt.Sprintf("umount/custom/%s/%s", poolName, volumeName)
+}
+
+func getImageCreateLockID(poolName string, fingerprint string) string {
+	return fmt.Sprintf("create/image/%s/%s", poolName, fingerprint)
+}
+
+// LockPoolMount creates a lock on the given pool for a mount operation,
+// and returns an unlock function which is to be called by the caller.
+func LockPoolMount(poolName string) func() {
```
Comment (tomponline) [inline lxd/storage/lock.go]: @monstermunchkin do these lock functions need to be exported (I only had a cursory glance, and couldnt see them being used outside of storage package).

File: lxd/storage/storage.go [hunk 15]
```diff
@@ -4,9 +4,93 @@ import (
...
+
+	UsesThinpool() bool
+	GetBlockFilesystem() string
+}
+
+// Driver represents a storage backend.
+type Driver interface {
+	GetVersion() string
+
+	StoragePoolCheck() error
```
Comment (tomponline) [inline lxd/storage/storage.go]: @monstermunchkin as these functions are part of an interface in the storage package, you could remove the "Storage" prefix from these function definitions to reduce "stuttering", like the Volume ones below.

**Comment by tomponline** [comment]
@monstermunchkin I've had a look through this PR to get some familiarity with the storage layer in LXD. I can see you've created 2 interfaces in the code, one for a sealed interface that is an unexported driver with common init, and the other an exported Driver interface with public functions.

I was interested in how you had linked your dir storage implementation into the existing LXD codebase, did you do it in the storage.go file, as there was a lot of code added there, which was a bit more than I was expecting, was this required because it relies on the `container` interface in the `main` package? If so, then you could add an `instance` interface to the storage package with only the functions the storage layer needs, this may allow you to then move more of that code into the storage package.

**Comment by tomponline** [comment]
@monstermunchkin I've re-run the tests and there is an issue with rsync by the looks of it:

```
 change_dir \"/lxc-ci/build/tmp.HhGhpXuQS5/go/src/github.com/lxc/lxd/test/tmp.xjC/BDG4G9IGM/storage-pools/lxdtest-BDG4G9IGM-dir/custom-snapshots/vol1/snap0\" failed: No such file or directory 
```

File: lxd/storage/dir.go (hunk 9 above)
Comment (monstermunchkin) [reply lxd/storage/dir.go]: The bool indicates whether the function actually mounted to volume. If the function returns `false`, it doesn't automatically mean there was an error. I don't think we should change the return values.

Comment (tomponline) [reply lxd/storage/dir.go]: @monstermunchkin ah, I see, so it could already be mounted? Do you know what scenario this would happen in?

File: lxd/storage/dir_utils.go (hunk 10 above)
Comment (monstermunchkin) [reply lxd/storage/dir_utils.go]: Interestingly, this function is not actually an internal function. It's used in `lxd/api_internal.go`. It therefore needs to be exported. Perhaps we should think of renaming this function :)

File: lxd/storage/common.go (hunk 11 above)
Comment (monstermunchkin) [reply lxd/storage/common.go]: It is specific to LVM, yes. However, I thought it'd be cleaner to add it to the common code instead of all other drivers just returning `false` in this function given that no LVM-specific code is called.

File: lxd/storage/storage.go (hunk 15 above)
Comment (monstermunchkin) [reply lxd/storage/storage.go]: Makes sense.

File: lxd/storage/common.go (hunk 11 above)
Comment (tomponline) [reply lxd/storage/common.go]: @monstermunchkin I see, in the current code base I can only see an unexported `usesThinpool` (https://github.com/lxc/lxd/search?q=usesThinpool&unscoped_q=usesThinpool) used only by LVM, so maybe this doesn't need to be part of the interface at all, and can just be part of the unexported implementation of LVM?

File: lxd/storage/dir.go (hunk 9 above)
Comment (monstermunchkin) [reply lxd/storage/dir.go]: Yes, it could already be mounted but I don't know what scenario this would happen in. I guess it's called sometimes just to makes sure something's really mounted.

File: lxd/storage/common.go (hunk 11 above)
Comment (monstermunchkin) [reply lxd/storage/common.go]: I just checked in my other branch, and the function is used in `lxd/storage.go` when creating a container. Maybe we can work around this, and remove the function from the interface.

Comment (tomponline) [reply lxd/storage/common.go]: @monstermunchkin yes I think that would be good, especially as in current master its only used in LVM related places. Do you remember why you used it in lxd/storage.go in your other branch?

**Comment by stgraber** [comment]
Just did a simple rebase + update to have this build against current master before I take a deeper look at how we can re-structure our storage to fit our current needs.

**Comment by stgraber** [comment]
Spent some time thinking through what I think we need to have things be nice and clean.

What I came up with is a `Pool` interface which would be held by the `storage` package.
This interface represents the operations that we expect the main components of the LXD codebase to drive as they use resources out of a storage pool.

I expect there to be three functions to get such a Pool:
```go
func CreatePool(state *state.State, pool api.StoragePool) (Pool, error)
func GetPoolByName(state *state.State, name string) (Pool, error)
func GetPoolByID(state *state.Staste, id int64) (Pool, error)
```

The interface itself would look something like:
```go
type Pool interface {
    // Internal
    DaemonState() *state.State

    // Pool
    ID() int64
    Name() string
    Driver() *Driver

    GetResources() (*api.ResourcesStoragePool, error)
    Delete(op *operations.Operation) error

    Mount() (bool, error)
    Unmount() (bool, error)

    // Instances
    CreateInstance(i Instance, op *operations.Operation) error
    CreateInstanceFromBackup(i Instance, sourcePath string, op *operations.Operation) error
    CreateInstanceFromCopy(i Instance, src Instance, snapshots bool, op *operations.Operation) error
    CreateInstanceFromImage(i Instance, fingerprint string, op *operations.Operation) error
    CreateInstanceFromMigration(i Instance, conn *websocket.Conn, args MigrationSinkArgs, op *operations.Operation) error
    RenameInstance(i Instance, newName string, op *operations.Operation) error
    DeleteInstance(i Instance, op *operations.Operation) error

    MigrateInstance(i Instance, snapshots bool, args MigrationSourceArgs) (MigrationStorageSourceDriver, error)
    RefreshInstance(i Instance, src Instance, snapshots bool, op *operations.Operation) error
    BackupInstance(i Instance, targetPath string, optimized bool, snapshots bool, op *operations.Operation) error

    GetInstanceUsage(i Instance) (uint64, error)
    SetInstanceQuota(i Instance, quota uint64) error

    MountInstance(i Instance) (bool, error)
    UnmountInstance(i Instance) (bool, error)
    GetInstanceDisk(i Instance) (string, string, error)

    // Instance snapshots
    CreateInstanceSnapshot(i Instance, name string, op *operations.Operation) error
    RenameInstanceSnapshot(i Instance, newName string, op *operations.Operation) error
    DeleteInstanceSnapshot(i Instance, op *operations.Operation) error

    RestoreInstanceSnapshot(i Instance, op *operations.Operation) error

    MountInstanceSnapshot(i Instance) (bool, error)
    UnmountInstanceSnapshot(i Instance) (bool, error)

    // Images
    CreateImage(img api.Image, op *operations.Operation) error
    DeleteImage(img api.Image, op *operations.Operation) error

    // Custom volumes
    CreateCustomVolume(vol api.StorageVolume, op *operations.Operation) error
    CreateCustomVolumeFromCopy(vol api.StorageVolume, src api.StorageVolume, snapshots bool, op *operations.Operation) error
    CreateCustomVolumeFromMigration(vol api.StorageVolume, conn *websocket.Conn, args MigrationSinkArgs, op *operations.Operation) error
    RenameCustomVolume(vol api.StorageVolume, newName string, op *operations.Operation) error
    DeleteCustomVolume(vol api.StorageVolume, op *operations.Operation) error

    MigrateCustomVolume(vol api.StorageVolume, snapshots bool, args MigrationSourceArgs) (MigrationStorageSourceDriver, error)

    GetCustomVolumeUsage(vol api.StorageVolume) (uint64, error)
    SetCustomVolumeQuota(vol api.StorageVolume, quota uint64) error

    MountCustomVolume(vol api.StorageVolume) (bool, error)
    UnmountCustomVolume(vol api.StorageVolume) (bool, error)

    // Custom volume snapshots
    CreateCustomVolumeSnapshot(vol api.StorageVolume, name string, op *operations.Operation) error
    RenameCustomVolumeSnapshot(vol api.StorageVolume, newName string, op *operations.Operation) error
    DeleteCustomVolumeSnapshot(vol api.StorageVolume, op *operations.Operation) error
}
```

I tried to keep things as consistent as possible and cleanup a lot of the mess we have with our current interface. With this, we can now have one struct per pool and store/share references to it as we see fit, the only internal state kept by the struct will be things like the remote credentials, filesystem paths, ... which can't be altered without deleting and re-creating the pool.

As originally planned, there will be two implementations of that interface:
 - `defaultBackend`
 - `mockBackend`

Note that those will not be exported types but kept internal to the `storage` packaged. To enable mock mode, an exported variable will be used (`storage.MockMode`), when flipped to true, the three functions described above will start returning a `mockBackend` instance rather than a `defaultBackend` one.

My next step is to poke at the `Driver` interface and flesh it out with the functions we expect to need to implement the above. Looking at what @monstermunchkin put in his version of that work.

After that, I plan to add the new interfaces in the `storage` package, write the loader functions, implement the mock backend and start fleshing out the default backend, likely initially driving a minimal version of the `dir` driver. This will need a lot of internal reshuffling to untie a lot of DB/storage relationships we've got today, but the result should be a useful package we can import wherever we need storage actions.

@monstermunchkin @tomponline if you happen to be lurking, comments welcome :)

**Comment by stgraber** [comment]
Closing in favor of #6285 and so my brain hurts less when looking at the PR list.
Please don't delete your branch though, we still want to pull some of the code from it!
